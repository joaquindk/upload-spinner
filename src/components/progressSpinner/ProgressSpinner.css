@keyframes rotating {
  from {
    transform: rotate(0deg);
  }
  to {
    transform: rotate(360deg);
  }
}
.spinnerWrapper {
  position: relative;
}

.spinning path:nth-child(2) {
  animation: rotating 2s linear infinite;
  will-change: transform;
}

.textWrapper {
  position: absolute;
  display: grid;
  font-weight: 600;
  grid-template-areas: 'number percentage' 'number .';
  grid-template-columns: auto auto;
  grid-column-gap: 4px;
  grid-template-rows: 50% 50%;
  top: 50%;
  left: 50%;
  transform: translate3d(-50%, -50%, 0);
}

.progressValue {
  grid-area: number;
}

.percentage {
  display: flex;
  justify-content: center;
  align-items: center;
  grid-area: percentage;
}

.videoWrapper {
  width: 100%;
  height: 100%;
  display: flex;
  justify-content: center;
  align-items: center;
}
