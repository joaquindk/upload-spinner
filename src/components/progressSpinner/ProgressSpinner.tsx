import React from 'react'
import cx from 'classnames'
import { Group } from '@vx/group'
import { Arc } from '@vx/shape'
import { ParentSize } from '@vx/responsive'
import styles from './ProgressSpinner.css'

/** Types of segments shown on the progress spinner. */
const SegmentTypes = Object.freeze({
  PROGRESS: 'PROGRESS',
  REMAINING: 'REMAINING'
})

type SegmentKeys = keyof typeof SegmentTypes;
type SegmentTypeValues = typeof SegmentTypes[SegmentKeys]

type ProgressEntry = {
  type: SegmentTypeValues,
  value: number
}

type Props = {
  /** Current progress. */
  progress: number,
  /** Total progress for which the task is considered completed. */
  total: number,
  /** Color used for the segment showing the currently completed progress. */
  progressColor: string,
  /** Color used for the segment showing the ramaining progress. */
  remainingColor: string,
  /** Color used for the background. */
  backgroundColor: string,
  /** Width of the circle stroke. */
  strokeWidth: number
  /** Font size for the progress text. */
  fontSizePx: number
  /** True if the progress arc should spin. */
  spin: boolean
}

/**
 * Component used to show the progress of a task. Given a 'progress' progress count
 * and a 'total' count, it will calculate the progress as a percentage and show a spinner
 * made up of 2 segments:
 *  1. A segment to showing current progress, shown in a color defined by the color prop.
 *  2. A placeholder segment showing the remaining progress. The current progress overlaps this segment which remain
 * on the background.
 * The current progress segment rotates clockwise until the task completes, if the spin prop is true.
 * After that a task completed state is shown with a video.
 */
const ProgressSpinner = ({ progress, total, progressColor, remainingColor, backgroundColor, strokeWidth, fontSizePx, spin }: Props) => {
  const progressRatio = progress / total
  const progressRad = progressRatio * 2 * Math.PI

  const renderArc = ({ endAngle, color, radius }: {endAngle: number, color: string, radius: number}) => (
    <Arc
      data={{}}
      outerRadius={radius}
      innerRadius={radius - strokeWidth}
      cornerRadius={strokeWidth}
      startAngle={0} endAngle={endAngle} fill={color}/>
  )

  return (
    <ParentSize>
      {({ width, height }) => {
        const centerY = height / 2
        const centerX = width / 2
        const radius = Math.min(width, height) / 2

        if (progressRatio >= 1) {
          return (
            <div className={styles.videoWrapper}>
              <video
                src="https://prod-cdn.wetransfer.net/packs/media/transfer_window/transfer_completed-3-5dbf2f01.mp4"
                autoPlay
                loop
                style={{ width: `${radius * 2}px` }}/>
            </div>
          )
        }

        return (
          <div className={cx(styles.spinnerWrapper, { [styles.spinning]: !!progress && spin })}>
            <svg width={width} height={height}>
              <rect rx={14} width={width} height={height} fill={backgroundColor} />
              <Group top={centerY} left={centerX}>
                {renderArc({ endAngle: 2 * Math.PI, color: remainingColor, radius })}
                {renderArc({ endAngle: progressRad, color: progressColor, radius })}
              </Group>
            </svg>
            <div className={styles.textWrapper} style={{ fontSize: `${fontSizePx}px`, lineHeight: `${fontSizePx}px` }}>
              <div className={styles.progressValue}>{`${Math.floor(progressRatio * 100)}`}</div>
              <div className={styles.percentage} style={{ fontSize: `${fontSizePx * 0.4}px`, color: remainingColor }}>%</div>
            </div>
          </div>
        )
      }}
    </ParentSize>)
}

ProgressSpinner.defaultProps = {
  progressColor: '#3687EF',
  remainingColor: '#E1E6E9',
  strokeWidth: 8,
  backgroundColor: 'white',
  fontSizePx: 72,
  spin: true
}

export default ProgressSpinner
