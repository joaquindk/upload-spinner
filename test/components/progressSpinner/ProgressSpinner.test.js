import React from 'react'
import ProgressSpinner from 'src/components/progressSpinner'
import { render } from '@testing-library/react'

// Mocks the ParentSize component to return a 200x200 size
jest.mock('@vx/responsive', () => ({
  ...jest.requireActual('@vx/responsive'),
  ParentSize: ({ children }) => children({ width: 200, height: 200 })
}))

describe('ProgressSpinner', () => {
  describe('with zero progress.', () => {
    it('should render a progress spinner with no progress', () => {
      const { container } = render(<ProgressSpinner progress={0} total={100} />)

      expect(container).toMatchSnapshot()
      expect(container.querySelectorAll('path').length).toBe(2)
      expect(container.querySelector('.progressValue').innerHTML).toBe('0')
    })

    it('should ignore the spin property and not make the spinner spin when the progress is zero.', () => {
      const { container } = render(<ProgressSpinner progress={0} total={100} spin />)

      expect(container.querySelectorAll('.spinning').length).toBe(0)
    })
  })

  describe('with some progress without having completed ', () => {
    it('should render a progress spinner with some progress that has not yet comleted', () => {
      const { container } = render(<ProgressSpinner progress={20} total={100} />)

      expect(container).toMatchSnapshot()
      expect(container.querySelectorAll('path').length).toBe(2)
      expect(container.querySelector('.progressValue').innerHTML).toBe('20')
      expect(container.querySelectorAll('.spinning').length).toBe(1)
    })

    it('should not make the progress segment spin if the spin prop is false', () => {
      const { container } = render(<ProgressSpinner progress={20} total={100} spin={false} />)

      expect(container).toMatchSnapshot()
      expect(container.querySelectorAll('.spinning').length).toBe(0)
    })

    it('should correctly apply each of the style properties passed to the spinner', () => {
      const componentProps = {
        progressColor: 'red',
        remainingColor: 'yellow',
        backgroundColor: 'green',
        fontSizePx: 20
      }
      const { container } = render(
        <ProgressSpinner progress={20} total={100} {...componentProps} />
      )

      expect(container.querySelectorAll('path')[1].getAttribute('fill')).toBe(
        componentProps.progressColor
      )
      expect(container.querySelectorAll('path')[0].getAttribute('fill')).toBe(
        componentProps.remainingColor
      )
      expect(container.querySelector('rect').getAttribute('fill')).toBe(
        componentProps.backgroundColor
      )
      expect(container.querySelector('.textWrapper').getAttribute('style')).toEqual(
        `font-size: ${componentProps.fontSizePx}px; line-height: ${componentProps.fontSizePx}px;`
      )
    })
  })

  describe('with finished progress', () => {
    it('should render a completed state video with auto play and loop', () => {
      const { container } = render(<ProgressSpinner progress={100} total={100} />)

      expect(container.querySelectorAll('svg').length).toBe(0)
      expect(container.querySelectorAll('video').length).toBe(1)
      expect(container).toMatchSnapshot()
    })
  })
})
