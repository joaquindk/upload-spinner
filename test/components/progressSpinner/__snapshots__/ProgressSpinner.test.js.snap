// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProgressSpinner with finished progress should render a completed state video with auto play and loop 1`] = `
<div>
  <div
    class="videoWrapper"
  >
    <video
      autoplay=""
      loop=""
      src="https://prod-cdn.wetransfer.net/packs/media/transfer_window/transfer_completed-3-5dbf2f01.mp4"
      style="width: 200px;"
    />
  </div>
</div>
`;

exports[`ProgressSpinner with some progress without having completed  should not make the progress segment spin if the spin prop is false 1`] = `
<div>
  <div
    class="spinnerWrapper"
  >
    <svg
      height="200"
      width="200"
    >
      <rect
        fill="white"
        height="200"
        rx="14"
        width="200"
      />
      <g
        class="vx-group"
        transform="translate(100, 100)"
      >
        <path
          class="vx-arc"
          d="M6.123233995736766e-15,-100A100,100,0,1,1,-6.123233995736766e-15,100A100,100,0,1,1,6.123233995736766e-15,-100M-1.6900125828233473e-14,-92A92,92,0,1,0,1.6900125828233473e-14,92A92,92,0,1,0,-1.6900125828233473e-14,-92Z"
          fill="#E1E6E9"
        />
        <path
          class="vx-arc"
          d="M3.552713678800501e-15,-95.91663046625439A4,4,0,0,1,4.1666666666666705,-99.91315673568165A100,100,0,0,1,93.73548796719345,-34.83759887753652A4,4,0,0,1,91.22213642600549,-29.63986885725444L91.22213642600549,-29.63986885725444A4,4,0,0,1,86.23664892981796,-32.0505909673336A92,92,0,0,0,3.833333333333333,-91.92010419682713A4,4,0,0,1,3.552713678800501e-15,-95.91663046625439Z"
          fill="#3687EF"
        />
      </g>
    </svg>
    <div
      class="textWrapper"
      style="font-size: 72px; line-height: 72px;"
    >
      <div
        class="progressValue"
      >
        20
      </div>
      <div
        class="percentage"
        style="font-size: 28.8px; color: rgb(225, 230, 233);"
      >
        %
      </div>
    </div>
  </div>
</div>
`;

exports[`ProgressSpinner with some progress without having completed  should render a progress spinner with some progress that has not yet comleted 1`] = `
<div>
  <div
    class="spinnerWrapper spinning"
  >
    <svg
      height="200"
      width="200"
    >
      <rect
        fill="white"
        height="200"
        rx="14"
        width="200"
      />
      <g
        class="vx-group"
        transform="translate(100, 100)"
      >
        <path
          class="vx-arc"
          d="M6.123233995736766e-15,-100A100,100,0,1,1,-6.123233995736766e-15,100A100,100,0,1,1,6.123233995736766e-15,-100M-1.6900125828233473e-14,-92A92,92,0,1,0,1.6900125828233473e-14,92A92,92,0,1,0,-1.6900125828233473e-14,-92Z"
          fill="#E1E6E9"
        />
        <path
          class="vx-arc"
          d="M3.552713678800501e-15,-95.91663046625439A4,4,0,0,1,4.1666666666666705,-99.91315673568165A100,100,0,0,1,93.73548796719345,-34.83759887753652A4,4,0,0,1,91.22213642600549,-29.63986885725444L91.22213642600549,-29.63986885725444A4,4,0,0,1,86.23664892981796,-32.0505909673336A92,92,0,0,0,3.833333333333333,-91.92010419682713A4,4,0,0,1,3.552713678800501e-15,-95.91663046625439Z"
          fill="#3687EF"
        />
      </g>
    </svg>
    <div
      class="textWrapper"
      style="font-size: 72px; line-height: 72px;"
    >
      <div
        class="progressValue"
      >
        20
      </div>
      <div
        class="percentage"
        style="font-size: 28.8px; color: rgb(225, 230, 233);"
      >
        %
      </div>
    </div>
  </div>
</div>
`;

exports[`ProgressSpinner with zero progress. should render a progress spinner with no progress 1`] = `
<div>
  <div
    class="spinnerWrapper"
  >
    <svg
      height="200"
      width="200"
    >
      <rect
        fill="white"
        height="200"
        rx="14"
        width="200"
      />
      <g
        class="vx-group"
        transform="translate(100, 100)"
      >
        <path
          class="vx-arc"
          d="M6.123233995736766e-15,-100A100,100,0,1,1,-6.123233995736766e-15,100A100,100,0,1,1,6.123233995736766e-15,-100M-1.6900125828233473e-14,-92A92,92,0,1,0,1.6900125828233473e-14,92A92,92,0,1,0,-1.6900125828233473e-14,-92Z"
          fill="#E1E6E9"
        />
        <path
          class="vx-arc"
          d="M6.123233995736766e-15,-100L5.633375276077825e-15,-92Z"
          fill="#3687EF"
        />
      </g>
    </svg>
    <div
      class="textWrapper"
      style="font-size: 72px; line-height: 72px;"
    >
      <div
        class="progressValue"
      >
        0
      </div>
      <div
        class="percentage"
        style="font-size: 28.8px; color: rgb(225, 230, 233);"
      >
        %
      </div>
    </div>
  </div>
</div>
`;
